File: framework/adapters.py
from __future__ import annotations

import base64
import os
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, List, Optional, Tuple

try:
    from appium import webdriver
    from appium.options.android import UiAutomator2Options
    from appium.options.ios import XCUITestOptions
    from appium.webdriver.common.appiumby import AppiumBy
    from selenium.webdriver.common.by import By
except Exception:  # pragma: no cover
    webdriver = None
    UiAutomator2Options = None
    XCUITestOptions = None
    AppiumBy = None
    By = None


DEFAULT_SERVER_URL = "http://127.0.0.1:4723"
DEFAULT_ANDROID_APP = "com.example.app"
DEFAULT_IOS_BUNDLE = "com.example.app"
DEFAULT_ANDROID_DEVICE = "emulator-5554"
DEFAULT_IOS_DEVICE = "iPhone 17 Pro"

# 1x1 transparent PNG
_MINI_PNG = base64.b64decode(
    "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO7Z0x8AAAAASUVORK5CYII="
)

ACCESSIBILITY_ID = getattr(AppiumBy, "ACCESSIBILITY_ID", "accessibility id")
XPATH = getattr(By, "XPATH", "xpath")
ELEMENT_ID = getattr(By, "ID", "id")
ANDROID_UIAUTOMATOR = getattr(AppiumBy, "ANDROID_UIAUTOMATOR", "-android uiautomator")
IOS_PREDICATE = getattr(AppiumBy, "IOS_PREDICATE", "-ios predicate string")
IOS_CLASS_CHAIN = getattr(AppiumBy, "IOS_CLASS_CHAIN", "-ios class chain")


@dataclass
class AdapterCapability:
    platform: str
    commands: List[str]


class PlatformAdapter(ABC):
    @abstractmethod
    def capability(self) -> AdapterCapability:
        raise NotImplementedError

    @abstractmethod
    def navigate(self, args: Dict) -> Dict:
        raise NotImplementedError

    @abstractmethod
    def tap(self, args: Dict) -> Dict:
        raise NotImplementedError

    @abstractmethod
    def act(self, args: Dict) -> Dict:
        raise NotImplementedError

    @abstractmethod
    def screenshot(self, args: Dict) -> Dict:
        raise NotImplementedError

    @abstractmethod
    def wait(self, args: Dict) -> Dict:
        raise NotImplementedError

    @abstractmethod
    def source(self, args: Dict) -> Dict:
        raise NotImplementedError

    @abstractmethod
    def exists(self, target: str) -> bool:
        raise NotImplementedError

    @abstractmethod
    def close(self) -> None:
        raise NotImplementedError


def _parse_target(target: str) -> Tuple[str, str]:
    if target.startswith("text="):
        value = target.split("=", 1)[1]
        xpath = (
            f"//*[contains(@text, '{value}') or contains(@content-desc, '{value}') "
            f"or contains(@label, '{value}') or contains(@name, '{value}') or contains(@value, '{value}')]"
        )
        return XPATH, xpath
    if target.startswith("id="):
        value = target.split("=", 1)[1]
        return ELEMENT_ID, value
    if target.startswith("xpath="):
        value = target.split("=", 1)[1]
        return XPATH, value
    if target.startswith("uiautomator="):
        value = target.split("=", 1)[1]
        return ANDROID_UIAUTOMATOR, value
    if target.startswith("predicate="):
        value = target.split("=", 1)[1]
        return IOS_PREDICATE, value
    if target.startswith("classchain="):
        value = target.split("=", 1)[1]
        return IOS_CLASS_CHAIN, value
    if target.startswith("accessibility="):
        value = target.split("=", 1)[1]
        return ACCESSIBILITY_ID, value
    return ACCESSIBILITY_ID, target


def _resolve_tap_mode(args: Dict) -> str:
    has_target = "target" in args and args["target"] is not None
    has_x = "x" in args and args["x"] is not None
    has_y = "y" in args and args["y"] is not None

    if has_target and (has_x or has_y):
        raise ValueError("tap cannot mix target with x/y coordinates")
    if has_target:
        return "target"
    if has_x and has_y:
        return "coordinates"
    if has_x != has_y:
        raise ValueError("tap coordinate mode requires both x and y")
    raise ValueError("tap requires target or x/y coordinates")


def _png_dimensions(path: Path) -> Tuple[int | None, int | None]:
    try:
        with path.open("rb") as f:
            header = f.read(24)
    except OSError:
        return None, None

    if len(header) < 24 or header[:8] != b"\x89PNG\r\n\x1a\n":
        return None, None
    return int.from_bytes(header[16:20], "big"), int.from_bytes(header[20:24], "big")


def _env(preferred: str, legacy: str, default: str | None = None) -> str | None:
    return os.getenv(preferred) or os.getenv(legacy) or default


def _env_bool(preferred: str, legacy: str, default: bool = False) -> bool:
    raw = _env(preferred, legacy)
    if raw is None:
        return default
    return raw.strip().lower() in {"1", "true", "yes", "on"}


class RealAppiumAdapter(PlatformAdapter):
    def __init__(
        self,
        platform: str,
        server_url: str,
        device: Optional[str] = None,
        app_id: Optional[str] = None,
        attach_to_running: bool = False,
    ) -> None:
        self.platform = platform
        self.server_url = server_url
        self.device = device
        self.app_id = app_id
        self.attach_to_running = attach_to_running
        self.driver = self._create_driver()

    def _create_driver(self):
        if webdriver is None:
            raise RuntimeError("Appium/Selenium dependencies are missing. Install in runtime environment or use --mock.")

        attach_to_running = self.attach_to_running or _env_bool(
            "VISOR_ATTACH_TO_RUNNING",
            "PATF_ATTACH_TO_RUNNING",
            default=False,
        )

        if self.platform == "android":
            options = UiAutomator2Options()
            options.platform_name = "Android"
            options.automation_name = "UiAutomator2"
            options.udid = self.device or _env("VISOR_ANDROID_DEVICE", "PATF_ANDROID_DEVICE", DEFAULT_ANDROID_DEVICE)
            options.app_package = self.app_id or _env("VISOR_ANDROID_APP_PACKAGE", "PATF_ANDROID_APP_PACKAGE", DEFAULT_ANDROID_APP)
            options.app_activity = _env("VISOR_ANDROID_APP_ACTIVITY", "PATF_ANDROID_APP_ACTIVITY", ".MainActivity")
            options.new_command_timeout = 60
            if attach_to_running:
                options.set_capability("noReset", True)
                options.set_capability("fullReset", False)
                options.set_capability("autoLaunch", False)
                options.set_capability("dontStopAppOnReset", True)
            return webdriver.Remote(self.server_url, options=options)

        if self.platform == "ios":
            options = XCUITestOptions()
            options.platform_name = "iOS"
            options.automation_name = "XCUITest"
            options.device_name = self.device or _env("VISOR_IOS_DEVICE", "PATF_IOS_DEVICE", DEFAULT_IOS_DEVICE)
            options.bundle_id = self.app_id or _env("VISOR_IOS_BUNDLE_ID", "PATF_IOS_BUNDLE_ID", DEFAULT_IOS_BUNDLE)
            options.new_command_timeout = 60
            if attach_to_running:
                options.set_capability("noReset", True)
                options.set_capability("fullReset", False)
                options.set_capability("autoLaunch", False)
                options.set_capability("shouldTerminateApp", False)
                options.set_capability("forceAppLaunch", False)
            return webdriver.Remote(self.server_url, options=options)

        raise ValueError(f"Unsupported platform: {self.platform}")

    def capability(self) -> AdapterCapability:
        return AdapterCapability(platform=self.platform, commands=["navigate", "tap", "act", "screenshot", "wait", "source"])

    def navigate(self, args: Dict) -> Dict:
        to = args.get("to", "")
        if to:
            self.driver.get(to)
        return {"action": "navigate", "platform": self.platform, "args": {"to": to}}

    def tap(self, args: Dict) -> Dict:
        if _resolve_tap_mode(args) == "coordinates":
            x, y = self._resolve_coordinates(args)
            self._tap_point(x, y)
            return {
                "action": "tap",
                "platform": self.platform,
                "args": {"x": x, "y": y, "normalized": bool(args.get("normalized", False))},
            }

        target = args["target"]
        by, value = _parse_target(target)
        elem = self.driver.find_element(by, value)
        elem.click()
        return {"action": "tap", "platform": self.platform, "args": {"target": target}}

    def act(self, args: Dict) -> Dict:
        name = args.get("name")
        value = args.get("value", "")
        target = args.get("target")
        if name == "type" and target:
            by, sel = _parse_target(target)
            elem = self.driver.find_element(by, sel)
            elem.clear()
            elem.send_keys(value)
            return {"action": "act", "platform": self.platform, "args": {"name": name, "target": target, "value": value}}
        if name == "back":
            self.driver.back()
            return {"action": "act", "platform": self.platform, "args": {"name": name}}
        raise ValueError("Unsupported act operation; use --name type --target <selector> --value <text> or --name back")

    def screenshot(self, args: Dict) -> Dict:
        label = args.get("label", "capture")
        requested_path = args.get("path")
        file_path = Path(requested_path) if requested_path else Path(f"{label}.png")
        file_path.parent.mkdir(parents=True, exist_ok=True)
        ok = self.driver.get_screenshot_as_file(str(file_path))
        if not ok:
            raise RuntimeError("Appium failed to persist screenshot")
        width, height = _png_dimensions(file_path)
        return {
            "action": "screenshot",
            "platform": self.platform,
            "args": {"label": label, "file": file_path.name, "path": str(file_path), "width": width, "height": height},
        }

    def wait(self, args: Dict) -> Dict:
        ms = int(args.get("ms", 0))
        if ms < 0:
            raise ValueError("wait requires non-negative ms")
        time.sleep(ms / 1000)
        return {"action": "wait", "platform": self.platform, "args": {"ms": ms}}

    def source(self, args: Dict) -> Dict:
        label = args.get("label", "source")
        requested_path = args.get("path")
        file_path = Path(requested_path) if requested_path else Path(f"{label}.xml")
        file_path.parent.mkdir(parents=True, exist_ok=True)
        content = self.driver.page_source
        file_path.write_text(content, encoding="utf-8")
        return {
            "action": "source",
            "platform": self.platform,
            "args": {"label": label, "file": file_path.name, "path": str(file_path), "format": "xml", "bytes": file_path.stat().st_size},
        }

    def exists(self, target: str) -> bool:
        by, value = _parse_target(target)
        return len(self.driver.find_elements(by, value)) > 0

    def _resolve_coordinates(self, args: Dict) -> Tuple[int, int]:
        x = float(args["x"])
        y = float(args["y"])
        if args.get("normalized", False):
            size = self.driver.get_window_size()
            x = x * size["width"]
            y = y * size["height"]
        return round(x), round(y)

    def _tap_point(self, x: int, y: int) -> None:
        if self.platform == "android":
            self.driver.execute_script("mobile: clickGesture", {"x": x, "y": y})
            return
        if self.platform == "ios":
            self.driver.execute_script("mobile: tap", {"x": x, "y": y})
            return
        raise ValueError(f"Coordinate tap is unsupported for platform: {self.platform}")

    def close(self) -> None:
        if self.driver:
            self.driver.quit()


class MockAdapter(PlatformAdapter):
    def __init__(self, platform: str) -> None:
        self.platform = platform

    def capability(self) -> AdapterCapability:
        return AdapterCapability(platform=self.platform, commands=["navigate", "tap", "act", "screenshot", "wait", "source"])

    def navigate(self, args: Dict) -> Dict:
        return {"action": "navigate", "platform": self.platform, "args": args}

    def tap(self, args: Dict) -> Dict:
        _resolve_tap_mode(args)
        return {"action": "tap", "platform": self.platform, "args": args}

    def act(self, args: Dict) -> Dict:
        return {"action": "act", "platform": self.platform, "args": args}

    def screenshot(self, args: Dict) -> Dict:
        label = args.get("label", "capture")
        requested_path = args.get("path")
        file_path = Path(requested_path) if requested_path else Path(f"{label}.png")
        file_path.parent.mkdir(parents=True, exist_ok=True)
        file_path.write_bytes(_MINI_PNG)
        width, height = _png_dimensions(file_path)
        return {
            "action": "screenshot",
            "platform": self.platform,
            "args": {"label": label, "file": file_path.name, "path": str(file_path), "width": width, "height": height},
        }

    def wait(self, args: Dict) -> Dict:
        ms = int(args.get("ms", 0))
        if ms < 0:
            raise ValueError("wait requires non-negative ms")
        return {"action": "wait", "platform": self.platform, "args": {"ms": ms}}

    def source(self, args: Dict) -> Dict:
        label = args.get("label", "source")
        requested_path = args.get("path")
        file_path = Path(requested_path) if requested_path else Path(f"{label}.xml")
        file_path.parent.mkdir(parents=True, exist_ok=True)
        content = f'<hierarchy platform="{self.platform}"><node text="mock" /></hierarchy>\n'
        file_path.write_text(content, encoding="utf-8")
        return {
            "action": "source",
            "platform": self.platform,
            "args": {"label": label, "file": file_path.name, "path": str(file_path), "format": "xml", "bytes": file_path.stat().st_size},
        }

    def exists(self, target: str) -> bool:
        lowered = target.lower()
        return "missing" not in lowered and "not_found" not in lowered

    def close(self) -> None:
        return None


def get_adapter(
    platform: str,
    server_url: str = DEFAULT_SERVER_URL,
    device: Optional[str] = None,
    use_mock: bool = False,
    app_id: Optional[str] = None,
    attach_to_running: bool = False,
) -> PlatformAdapter:
    normalized = platform.lower()
    if use_mock:
        return MockAdapter(normalized)
    return RealAppiumAdapter(
        normalized,
        server_url=server_url,
        device=device,
        app_id=app_id,
        attach_to_running=attach_to_running,
    )

File: src/adapters.ts
import fs from 'node:fs';
import path from 'node:path';

import { remote } from 'webdriverio';

import type { AdapterCapability, Platform, PlatformAdapter } from './types.js';
import { errorMessage, parseServerUrl, sleep } from './utils.js';

export const DEFAULT_SERVER_URL = 'http://127.0.0.1:4723';
export const DEFAULT_ANDROID_APP = 'com.example.app';
export const DEFAULT_IOS_BUNDLE = 'com.example.app';
export const DEFAULT_ANDROID_DEVICE = 'emulator-5554';
export const DEFAULT_IOS_DEVICE = 'iPhone 17 Pro';

const MINI_PNG_BASE64 =
  'iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO7Z0x8AAAAASUVORK5CYII=';
const MINI_PNG = Buffer.from(MINI_PNG_BASE64, 'base64');

export const ACCESSIBILITY_ID = 'accessibility id';
export const XPATH = 'xpath';
export const ELEMENT_ID = 'id';
export const ANDROID_UIAUTOMATOR = '-android uiautomator';
export const IOS_PREDICATE = '-ios predicate string';
export const IOS_CLASS_CHAIN = '-ios class chain';

type TapMode = 'target' | 'coordinates';
type ScrollDirection = 'up' | 'down';
type RemoteSession = Awaited<ReturnType<typeof remote>>;

interface ParsedTarget {
  strategy: string;
  value: string;
  selector: string;
}

function env(preferred: string, legacy: string, defaultValue?: string): string | undefined {
  return process.env[preferred] ?? process.env[legacy] ?? defaultValue;
}

function envBool(preferred: string, legacy: string, defaultValue = false): boolean {
  const raw = env(preferred, legacy);
  if (raw === undefined) {
    return defaultValue;
  }

  return ['1', 'true', 'yes', 'on'].includes(raw.trim().toLowerCase());
}

function pngDimensions(filePath: string): { width: number | null; height: number | null } {
  try {
    const header = fs.readFileSync(filePath).subarray(0, 24);
    if (
      header.length < 24 ||
      !header.subarray(0, 8).equals(Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]))
    ) {
      return { width: null, height: null };
    }

    return {
      width: header.readUInt32BE(16),
      height: header.readUInt32BE(20)
    };
  } catch {
    return { width: null, height: null };
  }
}

export function formatDriverCreationError(
  platform: Platform,
  appId: string | undefined,
  attachToRunning: boolean,
  error: unknown
): string {
  const message = errorMessage(error);

  if (platform === 'ios' && message.includes('bundle identifier') && message.includes('unknown')) {
    const targetApp = appId ?? DEFAULT_IOS_BUNDLE;
    const attachHint = attachToRunning
      ? ' When using --attach, launch that app on the simulator/device first.'
      : '';
    return `Failed to create WebdriverIO Appium session: ${message}. On iOS, --app-id must be the exact installed bundle identifier for the target app (${targetApp}). Android package names do not carry over automatically.${attachHint}`;
  }

  return `Failed to create WebdriverIO Appium session: ${message}`;
}

export function parseTarget(target: string): [string, string] {
  if (target.startsWith('text=')) {
    const value = target.split('=', 2)[1] ?? '';
    const xpath = `//*[contains(@text, '${value}') or contains(@content-desc, '${value}') or contains(@label, '${value}') or contains(@name, '${value}') or contains(@value, '${value}')]`;
    return [XPATH, xpath];
  }

  if (target.startsWith('id=')) {
    return [ELEMENT_ID, target.split('=', 2)[1] ?? ''];
  }

  if (target.startsWith('xpath=')) {
    return [XPATH, target.split('=', 2)[1] ?? ''];
  }

  if (target.startsWith('uiautomator=')) {
    return [ANDROID_UIAUTOMATOR, target.split('=', 2)[1] ?? ''];
  }

  if (target.startsWith('predicate=')) {
    return [IOS_PREDICATE, target.split('=', 2)[1] ?? ''];
  }

  if (target.startsWith('classchain=')) {
    return [IOS_CLASS_CHAIN, target.split('=', 2)[1] ?? ''];
  }

  if (target.startsWith('accessibility=')) {
    return [ACCESSIBILITY_ID, target.split('=', 2)[1] ?? ''];
  }

  return [ACCESSIBILITY_ID, target];
}

function selectorForTarget(target: string): ParsedTarget {
  const [strategy, value] = parseTarget(target);

  if (strategy === ACCESSIBILITY_ID) {
    return { strategy, value, selector: `~${value}` };
  }

  if (strategy === XPATH) {
    return { strategy, value, selector: value };
  }

  if (strategy === ELEMENT_ID) {
    return { strategy, value, selector: `id=${value}` };
  }

  if (strategy === ANDROID_UIAUTOMATOR) {
    return { strategy, value, selector: `android=${value}` };
  }

  if (strategy === IOS_PREDICATE) {
    return { strategy, value, selector: `-ios predicate string:${value}` };
  }

  return { strategy, value, selector: `-ios class chain:${value}` };
}

export function resolveTapMode(args: Record<string, unknown>): TapMode {
  const hasTarget = Object.hasOwn(args, 'target') && args.target !== undefined && args.target !== null;
  const hasX = Object.hasOwn(args, 'x') && args.x !== undefined && args.x !== null;
  const hasY = Object.hasOwn(args, 'y') && args.y !== undefined && args.y !== null;

  if (hasTarget && (hasX || hasY)) {
    throw new Error('tap cannot mix target with x/y coordinates');
  }

  if (hasTarget) {
    return 'target';
  }

  if (hasX && hasY) {
    return 'coordinates';
  }

  if (hasX !== hasY) {
    throw new Error('tap coordinate mode requires both x and y');
  }

  throw new Error('tap requires target or x/y coordinates');
}

function resolveScrollOptions(
  args: Record<string, unknown>
): { direction: ScrollDirection; percent: number; gesturePercent: number } {
  const direction = typeof args.direction === 'string' ? args.direction.toLowerCase() : '';
  if (direction !== 'up' && direction !== 'down') {
    throw new Error("scroll requires args.direction to be 'up' or 'down'");
  }

  const rawPercent = args.percent ?? 70;
  const percent = Number(rawPercent);
  if (!Number.isFinite(percent) || percent < 1 || percent > 100) {
    throw new Error('scroll args.percent must be a number between 1 and 100');
  }

  return {
    direction,
    percent,
    gesturePercent: percent / 100
  };
}

export class RealAppiumAdapter implements PlatformAdapter {
  private readonly platform: Platform;
  private readonly serverUrl: string;
  private readonly device?: string;
  private readonly appId?: string;
  private readonly attachToRunning: boolean;
  private driver: RemoteSession | null = null;

  private constructor(
    platform: Platform,
    serverUrl: string,
    device?: string,
    appId?: string,
    attachToRunning = false
  ) {
    this.platform = platform;
    this.serverUrl = serverUrl;
    this.device = device;
    this.appId = appId;
    this.attachToRunning = attachToRunning;
  }

  static async create(
    platform: Platform,
    serverUrl: string,
    device?: string,
    appId?: string,
    attachToRunning = false
  ): Promise<RealAppiumAdapter> {
    const adapter = new RealAppiumAdapter(platform, serverUrl, device, appId, attachToRunning);
    adapter.driver = await adapter.createDriver();
    return adapter;
  }

  capability(): AdapterCapability {
    return {
      platform: this.platform,
      commands: ['navigate', 'tap', 'act', 'scroll', 'screenshot', 'wait', 'source']
    };
  }

  async navigate(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const to = String(args.to ?? '');
    if (to) {
      await (this.requireDriver() as any).url(to);
    }

    return {
      action: 'navigate',
      platform: this.platform,
      args: { to }
    };
  }

  async tap(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    if (resolveTapMode(args) === 'coordinates') {
      const { x, y } = await this.resolveCoordinates(args);
      await this.tapPoint(x, y);
      return {
        action: 'tap',
        platform: this.platform,
        args: {
          x,
          y,
          normalized: Boolean(args.normalized)
        }
      };
    }

    const target = String(args.target);
    const selector = selectorForTarget(target);
    const element = await this.requireDriver().$(selector.selector);
    await element.click();
    return {
      action: 'tap',
      platform: this.platform,
      args: { target }
    };
  }

  async act(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const name = String(args.name ?? '');
    const value = String(args.value ?? '');
    const target = typeof args.target === 'string' ? args.target : undefined;

    if (name === 'type' && target) {
      const selector = selectorForTarget(target);
      const element = await this.requireDriver().$(selector.selector);
      await element.clearValue();
      await element.addValue(value);
      return {
        action: 'act',
        platform: this.platform,
        args: { name, target, value }
      };
    }

    if (name === 'back') {
      await (this.requireDriver() as any).back();
      return {
        action: 'act',
        platform: this.platform,
        args: { name }
      };
    }

    throw new Error(
      'Unsupported act operation; use --name type --target <selector> --value <text> or --name back'
    );
  }

  async scroll(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const { direction, percent, gesturePercent } = resolveScrollOptions(args);
    await this.scrollViewport(direction, gesturePercent);

    return {
      action: 'scroll',
      platform: this.platform,
      args: { direction, percent }
    };
  }

  async screenshot(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const label = String(args.label ?? 'capture');
    const filePath = path.resolve(typeof args.path === 'string' ? args.path : `${label}.png`);
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    await (this.requireDriver() as any).saveScreenshot(filePath);
    const { width, height } = pngDimensions(filePath);

    return {
      action: 'screenshot',
      platform: this.platform,
      args: {
        label,
        file: path.basename(filePath),
        path: filePath,
        width,
        height
      }
    };
  }

  async wait(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const ms = Number(args.ms ?? 0);
    if (ms < 0) {
      throw new Error('wait requires non-negative ms');
    }

    await sleep(ms);
    return {
      action: 'wait',
      platform: this.platform,
      args: { ms }
    };
  }

  async source(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const label = String(args.label ?? 'source');
    const filePath = path.resolve(typeof args.path === 'string' ? args.path : `${label}.xml`);
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    const content = await (this.requireDriver() as any).getPageSource();
    fs.writeFileSync(filePath, content, 'utf8');

    return {
      action: 'source',
      platform: this.platform,
      args: {
        label,
        file: path.basename(filePath),
        path: filePath,
        format: 'xml',
        bytes: fs.statSync(filePath).size
      }
    };
  }

  async exists(target: string): Promise<boolean> {
    const selector = selectorForTarget(target);
    const elements = (await this.requireDriver().$$(selector.selector)) as unknown as Array<unknown>;
    return elements.length > 0;
  }

  async close(): Promise<void> {
    if (!this.driver) {
      return;
    }

    const currentDriver = this.driver;
    this.driver = null;
    await currentDriver.deleteSession();
  }

  private requireDriver(): RemoteSession {
    if (!this.driver) {
      throw new Error('Driver session is not initialized');
    }

    return this.driver;
  }

  private async createDriver(): Promise<RemoteSession> {
    const attachToRunning =
      this.attachToRunning ||
      envBool('VISOR_ATTACH_TO_RUNNING', 'PATF_ATTACH_TO_RUNNING', false);
    const server = parseServerUrl(this.serverUrl);
    const capabilities: Record<string, unknown> = {};

    if (this.platform === 'android') {
      capabilities.platformName = 'Android';
      capabilities['appium:automationName'] = 'UiAutomator2';
      capabilities['appium:udid'] =
        this.device ?? env('VISOR_ANDROID_DEVICE', 'PATF_ANDROID_DEVICE', DEFAULT_ANDROID_DEVICE);
      capabilities['appium:appPackage'] =
        this.appId ?? env('VISOR_ANDROID_APP_PACKAGE', 'PATF_ANDROID_APP_PACKAGE', DEFAULT_ANDROID_APP);
      capabilities['appium:appActivity'] = env(
        'VISOR_ANDROID_APP_ACTIVITY',
        'PATF_ANDROID_APP_ACTIVITY',
        '.MainActivity'
      );
      capabilities['appium:newCommandTimeout'] = 60;

      if (attachToRunning) {
        capabilities['appium:noReset'] = true;
        capabilities['appium:fullReset'] = false;
        capabilities['appium:autoLaunch'] = false;
        capabilities['appium:dontStopAppOnReset'] = true;
      }
    } else {
      capabilities.platformName = 'iOS';
      capabilities['appium:automationName'] = 'XCUITest';
      capabilities['appium:deviceName'] =
        this.device ?? env('VISOR_IOS_DEVICE', 'PATF_IOS_DEVICE', DEFAULT_IOS_DEVICE);
      capabilities['appium:bundleId'] =
        this.appId ?? env('VISOR_IOS_BUNDLE_ID', 'PATF_IOS_BUNDLE_ID', DEFAULT_IOS_BUNDLE);
      capabilities['appium:newCommandTimeout'] = 60;

      if (attachToRunning) {
        capabilities['appium:noReset'] = true;
        capabilities['appium:fullReset'] = false;
        capabilities['appium:autoLaunch'] = false;
        capabilities['appium:shouldTerminateApp'] = false;
        capabilities['appium:forceAppLaunch'] = false;
      }
    }

    try {
      return await remote({
        protocol: server.protocol,
        hostname: server.host,
        port: server.port,
        path: server.pathname,
        capabilities,
        logLevel: 'error'
      });
    } catch (error) {
      throw new Error(
        formatDriverCreationError(this.platform, this.appId, attachToRunning, error)
      );
    }
  }

  private async resolveCoordinates(args: Record<string, unknown>): Promise<{ x: number; y: number }> {
    let x = Number(args.x);
    let y = Number(args.y);
    if (Boolean(args.normalized)) {
      const size = await (this.requireDriver() as any).getWindowSize();
      x *= Number(size.width);
      y *= Number(size.height);
    }

    return {
      x: Math.round(x),
      y: Math.round(y)
    };
  }

  private async tapPoint(x: number, y: number): Promise<void> {
    const driver = this.requireDriver() as any;
    if (this.platform === 'android') {
      await driver.execute('mobile: clickGesture', [{ x, y }]);
      return;
    }

    if (this.platform === 'ios') {
      await driver.execute('mobile: tap', [{ x, y }]);
      return;
    }

    throw new Error(`Coordinate tap is unsupported for platform: ${this.platform}`);
  }

  private async scrollViewport(direction: ScrollDirection, gesturePercent: number): Promise<void> {
    const driver = this.requireDriver() as any;
    const size = await driver.getWindowSize();
    const left = Math.max(0, Math.round(size.width * 0.1));
    const top = Math.max(0, Math.round(size.height * 0.1));
    const width = Math.max(1, Math.round(size.width * 0.8));
    const height = Math.max(1, Math.round(size.height * 0.8));

    if (this.platform === 'android') {
      await driver.execute('mobile: scrollGesture', [
        { left, top, width, height, direction, percent: gesturePercent }
      ]);
      return;
    }

    if (this.platform === 'ios') {
      try {
        await driver.execute('mobile: scrollGesture', [
          { left, top, width, height, direction, percent: gesturePercent }
        ]);
        return;
      } catch {
        await this.swipeViewport(direction, gesturePercent, size.width, size.height);
        return;
      }
    }

    throw new Error(`Scroll is unsupported for platform: ${this.platform}`);
  }

  private async swipeViewport(
    direction: ScrollDirection,
    gesturePercent: number,
    viewportWidth: number,
    viewportHeight: number
  ): Promise<void> {
    const driver = this.requireDriver() as any;
    const x = Math.round(viewportWidth / 2);
    const lowY = Math.round(viewportHeight * 0.75);
    const highY = Math.round(viewportHeight * 0.25);
    const travel = Math.max(1, Math.round(viewportHeight * gesturePercent));
    const startY = direction === 'down' ? lowY : highY;
    const unclampedEndY = direction === 'down' ? startY - travel : startY + travel;
    const endY = Math.max(1, Math.min(viewportHeight - 1, unclampedEndY));

    await driver.performActions([
      {
        type: 'pointer',
        id: 'finger1',
        parameters: { pointerType: 'touch' },
        actions: [
          { type: 'pointerMove', duration: 0, x, y: startY },
          { type: 'pointerDown', button: 0 },
          { type: 'pause', duration: 150 },
          { type: 'pointerMove', duration: 400, x, y: endY },
          { type: 'pointerUp', button: 0 }
        ]
      }
    ]);
    await driver.releaseActions();
  }
}

export class MockAdapter implements PlatformAdapter {
  constructor(private readonly platform: Platform) {}

  capability(): AdapterCapability {
    return {
      platform: this.platform,
      commands: ['navigate', 'tap', 'act', 'scroll', 'screenshot', 'wait', 'source']
    };
  }

  async navigate(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    return { action: 'navigate', platform: this.platform, args };
  }

  async tap(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    resolveTapMode(args);
    return { action: 'tap', platform: this.platform, args };
  }

  async act(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    return { action: 'act', platform: this.platform, args };
  }

  async scroll(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const { direction, percent } = resolveScrollOptions(args);
    return {
      action: 'scroll',
      platform: this.platform,
      args: { direction, percent }
    };
  }

  async screenshot(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const label = String(args.label ?? 'capture');
    const filePath = path.resolve(typeof args.path === 'string' ? args.path : `${label}.png`);
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    fs.writeFileSync(filePath, MINI_PNG);
    const { width, height } = pngDimensions(filePath);
    return {
      action: 'screenshot',
      platform: this.platform,
      args: {
        label,
        file: path.basename(filePath),
        path: filePath,
        width,
        height
      }
    };
  }

  async wait(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const ms = Number(args.ms ?? 0);
    if (ms < 0) {
      throw new Error('wait requires non-negative ms');
    }

    return { action: 'wait', platform: this.platform, args: { ms } };
  }

  async source(args: Record<string, unknown>): Promise<Record<string, unknown>> {
    const label = String(args.label ?? 'source');
    const filePath = path.resolve(typeof args.path === 'string' ? args.path : `${label}.xml`);
    fs.mkdirSync(path.dirname(filePath), { recursive: true });
    const content = `<hierarchy platform="${this.platform}"><node text="mock" /></hierarchy>\n`;
    fs.writeFileSync(filePath, content, 'utf8');
    return {
      action: 'source',
      platform: this.platform,
      args: {
        label,
        file: path.basename(filePath),
        path: filePath,
        format: 'xml',
        bytes: fs.statSync(filePath).size
      }
    };
  }

  async exists(target: string): Promise<boolean> {
    const lowered = target.toLowerCase();
    return !lowered.includes('missing') && !lowered.includes('not_found');
  }

  async close(): Promise<void> {
    return;
  }
}

export async function getAdapter(
  platform: string,
  serverUrl = DEFAULT_SERVER_URL,
  device?: string,
  useMock = false,
  appId?: string,
  attachToRunning = false
): Promise<PlatformAdapter> {
  const normalized = platform.toLowerCase() as Platform;
  if (useMock) {
    return new MockAdapter(normalized);
  }

  return RealAppiumAdapter.create(normalized, serverUrl, device, appId, attachToRunning);
}
